### PR TITLE
Get-DbaDependency

### PR DESCRIPTION
**Pull request by ctrlbold**
@ConstantineK is looking into this and seems to have gotten it to work. 

My question was that "My shitty recursion doesn't work. Why?"

### CODE REVIEW

File: functions/Get-DbaDependency.ps1 [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+                
+                [System.Object]
+                $Parent
+            )
+            
+            Add-Member -InputObject $InputObject -Name Parent -Value $Parent -MemberType NoteProperty
+            Add-Member -InputObject $InputObject -Name Tier -Value $Tier -MemberType NoteProperty -PassThru
+            
+            if ($InputObject.HasChildNodes) { Read-DependencyTree -InputObject $InputObject.FirstChild -Tier ($Tier + 1) -Parent $InputObject }
+            if ($InputObject.NextSibling) { Read-DependencyTree -InputObject $InputObject.NextSibling -Tier $Tier -Parent $Parent }
```
Comment (ctrlbold) [inline functions/Get-DbaDependency.ps1]: There's the 🏆 !! beautiful!!

**Comment by ctrlbold** [comment]
Love the logic! It's gorgeous 👍 I added one little change.

I did have a couple issues - can we have an easy switch to include the object we send in the output? That way, it'll make it easier to copy it to another place.

image

In addition, it doesn't get the objects in order (which should happen if the person wants to script out). For the good ol Northwind database, I believe it includes everything it needs but it attempted to create a view with a dependency on a table that isn't created until lower down in the output.

**Comment by FriedrichWeinmann** [comment]
Alright, fixed the sorting order:
The thing is: We need a different sorting order when enumerating prerequisites of the object.
So I included an if/else that checks for that.

**Comment by FriedrichWeinmann** [comment]
For sending along the input, I would like to direct your attention to the "OriginalResource" property ^^

**Comment by ctrlbold** [comment]
hey @FriedrichWeinmann  - thanks for all of the time you've invested in this command, especially considering it's probably not even interesting to you lol. 

Here's where I'm at. I ran the following

$table = (Get-DbaDatabase -SqlInstance sql2012 Northwind).tables | Where Name -eq Customers
$table | Get-DbaDependency

That returned nothing, which made me realize that I would like to turn -Dependents to -Parents, sorry. 

Even so, I ran the following ` $table | Get-DbaDependency -Dependents -IncludeSelf  | Select  -ExpandProperty Script > tables.sql`

Then I 

create database test
go
use test
go

Then I ran the contents of tables.sql and got:

``
Msg 208, Level 16, State 1, Procedure Invoices, Line 10 [Batch Start Line 25]
Invalid object name 'Shippers'.
Msg 1767, Level 16, State 0, Line 64
Foreign key 'FK_CustomerCustomerDemo' references invalid table 'dbo.CustomerDemographics'.
Msg 1750, Level 16, State 1, Line 64
Could not create constraint or index. See previous errors.
Msg 208, Level 16, State 1, Procedure Customer and Suppliers by City, Line 5 [Batch Start Line 67]
Invalid object name 'Suppliers'.
Msg 208, Level 16, State 1, Procedure Orders Qry, Line 6 [Batch Start Line 75]
Invalid object name 'Orders'.
Msg 208, Level 16, State 1, Procedure Quarterly Orders, Line 3 [Batch Start Line 83]
Invalid object name 'Orders'.
Msg 208, Level 16, State 1, Procedure Sales Totals by Amount, Line 4 [Batch Start Line 89]
Invalid object name 'Orders'.
Msg 1767, Level 16, State 0, Line 119
Foreign key 'FK_Orders_Employees' references invalid table 'dbo.Employees'.
Msg 1750, Level 16, State 1, Line 119
Could not create constraint or index. See previous errors.
Msg 208, Level 16, State 1, Procedure Sales by Category, Line 4 [Batch Start Line 122]
Invalid object name 'Categories'.
Msg 1767, Level 16, State 0, Line 170
Foreign key 'FK_Order_Details_Products' references invalid table 'dbo.Products'.
Msg 1750, Level 16, State 1, Line 170
Could not create constraint or index. See previous errors.
Msg 208, Level 16, State 1, Procedure Summary of Sales by Year, Line 3 [Batch Start Line 199]
Invalid object name 'Order Subtotals'.
Msg 208, Level 16, State 1, Procedure Product Sales for 1997, Line 4 [Batch Start Line 206]
Invalid object name 'Categories'.
Msg 208, Level 16, State 1, Procedure Summary of Sales by Quarter, Line 3 [Batch Start Line 217]
Invalid object name 'Order Subtotals'.
Msg 208, Level 16, State 1, Procedure Category Sales for 1997, Line 3 [Batch Start Line 230]
Invalid object name 'Product Sales for 1997'.
Msg 208, Level 16, State 1, Procedure Order Details Extended, Line 5 [Batch Start Line 248]
Invalid object name 'Products'.
```

If this continues to cause issues, we can just show dependencies and support scripts later. I know you have a ton on your plate!

**Comment by ctrlbold** [comment]
Thanks @FriedrichWeinmann! 🌮s for lyfe!

**Pull request merged**